irodsConnector/session.py

```diff
@@ -320,13 +336,32 @@ def _write_pam_password(self):
         authentication file in obfuscated form.
 
         """
-        connection = self._session.pool.get_connection()
-        pam_passwords = self._session.pam_pw_negotiated
+        connection = self._irods_session.pool.get_connection()
+        pam_passwords = self._irods_session.pam_pw_negotiated
         if len(pam_passwords):
-            irods_auth_file = self._session.get_irods_password_file()
+            irods_auth_file = self._irods_session.get_irods_password_file()
             with open(irods_auth_file, 'w', encoding='utf-8') as authfd:
                 authfd.write(
                     irods.password_obfuscation.encode(pam_passwords[0]))
         else:
             logging.info('WARNING -- unable to cache obfuscated password locally')
         connection.release()
+
+    def connect(self):
```

Comment:
What would speak against putting the whole authentication workflow here? 
In case that there is already a valid connection we could
1) cleanup that connection and reinstantiate a new one
2) return the existing session

The attribute irods_session would then also return either a valid connection or None.

Reply:
> The attribute irods_session would then also return either a valid connection or None.  Give that our frontend logic: do not create a new session when there is an authenticated one is already implemented in the manager.py, I think we can dare to be more general here and create a new session as soon as someone asks for it through the function.

I believe the `irods_session` property getter is doing exactly this already.  If no session is established, `self._irods_session` remains `None` when it is returned.  It can certainly be cleaned up a bit, but this is essentially the current situation.

The workflow currently resides in mainly three methods (four if you count the `@irods_session.deleter`).  That is too much code for a single method.  Further modularizing it (splitting up and simplifying `_get_irods_session()`, for instance) and moving the bulk of the _getter_ code to `connect()` certainly cannot hurt.  I believe the `irods_session` property should still be used to house the actual `iRODSSession`.  I just cannot predict its form in this new configuration.

One proposal I have in addition to this is perhaps to simplfy the `try...except` clause around `self.conn.connect()` in `iBridges.py` and instead capture the errors within the session and return them to the `login_function` already parsed.  Then that functionality can be used outside the GUI.

Shall I take a crack at all this to see what I come up with?

Reply:
>  What would speak against putting the whole authentication workflow here?
> In case that there is already a valid connection we could
>
> 1. cleanup that connection and reinstantiate a new one
> 2. return the existing session

Looking back at this, how do we determine the choice between **1.** and **2.**?

Currently, we create an `iRODSSession` with the authentication tokens for the chosen environment.  That session remains until "Close session" is selected.  At that point, the **context** is _reset_ and a whole _new_ `IrodsConnector` and associated _new_ `iRODSSession` is created with the _new_ authentication tokens for the _new_ chosen environment.  The two parts of **1.** are in two different workflows!

Returning the existing, valid session (**2.**) is the job of the `irods_session` property regardless of what else it may or may not do.

As for a valid _connection_, that is the responsibility of the `iRODSSession`.  It has its own `Connection`s, connection `Pool`s, and linked logic to arbitrate a valid _connection_.  The only definitive method to check for a _connected_ session is to query for the `server_version` property.  There are also **idle** and **active** pools holding the actual `Connection`s.  I would not want to use those facts because I don't know what logic (current and future) governs them.

One thing I _can_ do, though it may be overkill (and have performance consequences), is to let the `irods_session` be set by the `connect()` method then check it's _validity_ (get the `server_version`) before returning the _valid_ session instance.  This would happen _every time_ `context.session.irods_session` is accessed ensuring:

> The attribute irods_session would then also return either a valid connection or None.

Will this satisfy?

Reply:
Please have a look ar PR #89. This is a pull request to this PR/branch which tackles this problem.

Reply:
I had already finished my updates and the main content difference is the `greetings_from_irods()` method.  That is a good idea, but where is it decided to show it or not?

The main functional difference is the validity test for the `irods_session`.  I believe that is important for your goal.

Also, is the removal of the SSL context safe?  I have never really tested it.

Lastly, where do we prefer to retrieve the session introspection values from?  The environment file, the session, or both?

Reply:
Perhaps it is best to add a verbosity flag to the iBridges configuration.  QUIET for no greeting or other messages, NORMAL for the greeting and informational messages, DEBUG for _everything_.  Errors will always be shown.

Reply:
> I had already finished my updates and the main content difference is the `greetings_from_irods()` method. That is a good idea, but where is it decided to show it or not?
> 
The connect function has a boolean `greet`. Default is `False` but one can set it to `True`.
> The main functional difference is the validity test for the `irods_session`. I believe that is important for your goal.
> 
If the connect function fails it will create an Exception which is caught in the front end and then translated to a more useful error message for the user.  The backend can also not fix missing or wrong rods parameters. So I would say, we do not need extra checking in the backend.
> Also, is the removal of the SSL context safe? I have never really tested it.
It is, it was really just a convenience for YODA users at a time where the SSL context was not yet communicated to the users.
> 
> Lastly, where do we prefer to retrieve the session introspection values from? The environment file, the session, or both?
That is a tricky one. In the session.py, I would tend to retrieve all parameters from the actual irods_session as it is instantiated and authenticated. This would also give us the chance for debugging to compare with the parameters in context which are provided by the config files, just to test whether all parameters are set correctly in the actual irods_session.

Reply:
> > I had already finished my updates and the main content difference is the `greetings_from_irods()` method. That is a good idea, but where is it decided to show it or not?
> 
> The connect function has a boolean `greet`. Default is `False` but one can set it to `True`.

What I meant was, does the user decide whether or not to see this?  How would the user indicate that?  I believe that a "verbose" setting would be the most generic way to handle it.  Python logging has a  number of log levels.  This is clearly an INFO level message.  I can update the logger to output `logging.info()` both to the log file and to standard out.  This allows only one call instead of separate print statements.  The logging messages to the GUI uses a completely different mechanism, so that will need to be dealt with differently.

> > Lastly, where do we prefer to retrieve the session introspection values from? The environment file, the session, or both?
> 
> That is a tricky one. In the session.py, I would tend to retrieve all parameters from the actual irods_session as it is instantiated and authenticated. This would also give us the chance for debugging to compare with the parameters in context which are provided by the config files, just to test whether all parameters are set correctly in the actual irods_session.

I don't think that is a useful exercise.  The PRC is so well developed that these values will likely _always_ be set correctly.  If they weren't, the session would have other noticeable problems.

I think the only reason to use the session values is to give an indication that there _is_ a session.  Using `server_version` is the best way to determine that.  It is just simpler to take the values from the environment where possible.  That's what I would recommend.